buildSrc/src/main/groovy/org/elasticsearch/gradle/VersionCollection.groovy

```diff
@@ -20,164 +20,336 @@
+                            break
+                        }
+                    }
+                    // caveat 0 - now dip back 1 version to get the last supported snapshot version of the line
+                    Version highestMinor = getHighestPreviousMinor(currentVersion.major)
+                    maintenanceBugfixSnapshot = removeAndReaddAsSnapshot(highestMinor)
+                }
+            }
         }
 
-        this.versions = Collections.unmodifiableList(versions)
+        this.versions = Collections.unmodifiableList(versionSet.toList())
     }
 
     /**
      * @return The list of versions read from the Version.java file
      */
     List<Version> getVersions() {
-        return Collections.unmodifiableList(versions)
+        return versions
     }
 
     /**
-     * @return The latest version in the Version.java file, which must be the current version of the system.
+     * Index compat supports 1 previous entire major version. For instance, any 6.x test for this would test all of 5 up to that 6.x version
+     *
+     * @return All earlier versions that should be tested for index BWC with the current version.
      */
-    Version getCurrentVersion() {
-        return versions[-1]
+    List<Version> getVersionsIndexCompatibleWithCurrent() {
+        int actualMajor = (currentVersion.major == 5 ? 2 : currentVersion.major - 1)
+        return versionSet
+            .tailSet(Version.fromString("${actualMajor}.0.0"))
+            .headSet(currentVersion)
+            .asList()
+    }
+
+    /**
+     * Ensures the types of snapshot are not null and are also in the index compat list
+     */
+    List<Version> getSnapshotVersionsIndexCompatibleWithCurrent() {
+        List<Version> compatSnapshots = []
+        List<Version> allCompatVersions = getVersionsIndexCompatibleWithCurrent()
+        if (allCompatVersions.contains(nextMinorSnapshot)) {
+            compatSnapshots.add(nextMinorSnapshot)
+        }
+        if (allCompatVersions.contains(stagedMinorSnapshot)) {
+            compatSnapshots.add(stagedMinorSnapshot)
+        }
+        if (allCompatVersions.contains(nextBugfixSnapshot)) {
+            compatSnapshots.add(nextBugfixSnapshot)
+        }
+        if (allCompatVersions.contains(maintenanceBugfixSnapshot)) {
+            compatSnapshots.add(maintenanceBugfixSnapshot)
+        }
+
+        return compatSnapshots
     }
 
     /**
-     * @return The snapshot at the end of the previous minor series in the current major series, or null if this is the first minor series.
+     * Wire compat supports the last minor of the previous major. For instance, any 6.x test would test 5.6 up to that 6.x version
+     *
+     * @return All earlier versions that should be tested for wire BWC with the current version.
      */
-    Version getBWCSnapshotForCurrentMajor() {
-        return getLastSnapshotWithMajor(currentVersion.major)
+    List<Version> getVersionsWireCompatibleWithCurrent() {
```

Comment:
I find these longer names actually harder to read. Can we go back to the old names (circa mid last year) so we could have eg `bwcVersions.indexCompatible` and `bwcVersions.wireCompatible`? Ie `getIndexCompatible()` and `getWireCompatible()` methods here.